### ISSUE TITLE
Feature Request: Show more than one day of appointments

### ISSUE THREAD

**Issue by FoolishExcalibur** [lfos/calcurse#152]
I was wondering if there was a possibility to show more than one day worth of appointments in the app. window? Like you wouldn't need to go to the next day to see what kind of appointment it is but rather have a bunch listed from the beginning?

**Comment by lfos** [comment]
This is not currently possible. Sounds like a useful feature, though.

**Comment by FoolishExcalibur** [comment]
The only currently to me known method to do so is by appending -d*insert high number* 
It shouldnt be all too difficult to implement it.

**Comment by Ierezell** [comment]
@FoolishExcalibur Yes right now you can do it in CLI but I was thinking the same as you....
Please, could it be possible to list all the appointement in the appointment window (or at least all for the month or week) ? 

Btw, great job it's really a nice program to use !

**Comment by lhca** [comment]
I am looking into this and have a prototype mostly working that diplays a week at a time.

Anyone willing to try it out and provide feedback? I will make it available as a branch in the cloned project https://github.com/lhca/calcurse. You will have to build from source. We can discuss which base to use.

**Comment by koreymacdougall** [comment]
@lhca - I'd be interested in trying this out.  I looked quickly through your branches but couldn't find the commit/changes.  Is it currently up?

**Comment by lhca** [comment]
@koreymacdougall - It's not there yet because it's not quite ready. When it is, I will have to port it from my own private branch (not public) to one that is accessible, either pu or master? Have you any preferences?

**Comment by koreymacdougall** [comment]
@lhca - it wouldn't make any difference to me which branch it's on.  I'd be testing it as an isolated source build, away from a system install, so it wouldn't conflict with the mainline master, if that's what you're thinking about.  

I'm really interested to see your implementation.  I've been meaning to write this patch myself for awhile, but haven't had the time.  Glad to see someone making progress on it!

**Comment by lhca** [comment]
@koreymacdougall - Here it is https://github.com/lhca/calcurse/tree/multiple-days (based on pu).

Basically, seven days are loaded into the appointments panel instead of just one. Hopefully, it's intuitively clear how to proceed from there.

Just a few remarks. Seven days are loaded, starting with the day currently selected in the calendar. You can scroll up and down the list of appointments as before, but now across days. However, the day selected _in the calendar_ remains the same (the one highlighted in red). You can insert, delete, edit, copy, paste etc. in the list without changing the calendar day. For a day without any appointments or events, a dummy event, `(none)`, is inserted, making the day selectable. Another lot of days is loaded by selecting another day in the calendar (use the `w` or `W` command to go a week forwards or backwards).

The number of days is fixed at the moment. You can change it by setting the variable `day_nb` at the top of the file `days.c`.

**Comment by koreymacdougall** [comment]
This is great!  I haven't had much time to look into it yet, been
pretty busy this week, but many thanks for making this available.
I'll dig into over the weekend.

Cheers

**Comment by lhca** [comment]
Updated branch https://github.com/lhca/calcurse/tree/multiple-days with improvements in the automatic selection in the appointments panel. Mostly not very visible.

**Comment by lhca** [comment]
> Updated branch https://github.com/lhca/calcurse/tree/multiple-days with improvements in the automatic selection in the appointments panel. Mostly not very visible.

Further update. When up/down movements in the appointments panel reach top or bottom, a new lot of days is loaded and the selected day in the calendar changes to the first day of the new lot.

**Comment by lhca** [comment]
> The number of days is fixed at the moment. You can change it by setting the variable `day_nb` at the top of the file `days.c`.

It is now a configuration variable `general.multipledays` that may be set between one and seven.

**Comment by lfos** [comment]
Is there anything special about the number 7, anything that prevents us from displaying even more days?

**Comment by lfos** [comment]
My laptop has a screen resolution of 2560x1440 and a full screen terminal displays ~80 lines by default. I could imagine showing much more than 7 low-volume days there. Not sure how useful that is but I do not see any reason to prevent the user from doing so either.

I'm excited to see this feature in action. Thanks a lot for working on it!

**Comment by lhca** [comment]
> 
> My laptop has a screen resolution of 2560x1440 and a full screen terminal displays ~80 lines by default. I could imagine showing much more than 7 low-volume days there.

Admitted, seven days may be too few. However, my point remains valid. There is no point in loading days that are not shown. Imagine a situation with 10 days visible on average. What is the difference in the APP panel between loading 10 days and 365 days? The scrollbar!

But another aspect becomes more distinct with a high number of days loaded: the interaction with the calendar. Presently the calendar is controlled by the selected day, the red date. With only one day in the APP panel, the selected day (in the calendar) is also the selected day here. But with multiple days in the APP panel there are two selected days to consider: the day in the calendar and the day in the appointments panel. So far I have chosen to make the selected day in the calendar the first of the loaded days. With five days loaded, the days in the APP panel are (almost always) also visible in the calendar. With a very high number of days loaded, the APP panel may display days none of which are visible in the calendar. Is that a problem? I don't know. Experience will show.

> Not sure how useful that is but I do not see any reason to prevent the user from doing so either.

Do you have a suggestion for an upper limit on the number of days loaded?

**Comment by lfos** [comment]
I don't see a reason to impose any upper limit at all unless there is any reason to do so. If the user chooses to pick an inconveniently large number, it is his own choice.

**Comment by Ierezell** [comment]
Working perfectly on ArchLinux with termite ! :+1: Thanks a lot ! 
The only inconvenient was uninstalling the previous one... 
Some files stayed there and I didn't get the upgrade until I removed by hand some other files.

Still a really nice improvement !

**Comment by lhca** [comment]
I uploaded several improvements yesterday. Among them are full weeks in the calendar for the first and last week of the month, week numbers to the left of each week and an optional empty line after each appointment (to squeeze more lines into the panel). The latter is set in the general options menu where you can also change the number of days loaded (currently limited to seven days).
 
> (Could you put just an if condition to remove the none from the pannel ? Or perhaps I can do a fork and propose it in some times...)

That's not so easy as you think. Without it you wouldn't be able to add anything to an empty day.

The focus (highlighting) moves among the appointments and events as you press the movement keys. Without something to highlight, you can't see where you are, and the program can't select an empty day (the day headings, the dates, are always highlighted and not selectable). The text `(none)` is translatable, though.

**Comment by lhca** [comment]
The number of empty lines between consecutive days has also been made configurable (0, 1 or 2).

Changes have been uploaded; note that the commits have been slightly rearranged.

**Comment by lhca** [comment]
Further updates - my final take on this feature.

The connection between the selected day in the calendar and the selected day in the appointments panel has improved so that the calendar follows the appointments panel when you move up and down among the days. The selected day is indicated by a pair of red square brackets (e.g. `[30]`) that do not hide the colouring of the date.

**Comment by koreymacdougall** [comment]
@lhca - nice work with this :+1:   Sorry it has taken so long to get you feedback; been sort of off-grid.

I really like the ability to see my week at a glance.  That has been the major missing feature for me w/ calcurse, and this implementation is a great step toward that.  I also think it's great that you can scroll vertically through appointments and have the selected day change.  That dual-mode navigation makes scheduling less cognitively demanding (i.e., you don't have to context-switch and tab over to the calendar).  

There seems to be a bug when using this within tmux (v. 0.8.7.4).  It seems like screen artifacts: the (none) text and the date sometimes don't update or clear from the screen properly.  I think it is specifically when the display reaches the end of the days that have appointments scheduled.  I then end up with repeated, empty days.  See screenshot, the 3 Feb 16's at the end.  It doesn't show up when I launch, but when I move back and forth between days a few times. 

![calcurse_multiple_days](https://user-images.githubusercontent.com/22301892/52566586-1e892100-2dd8-11e9-9add-34919541990f.png)

Again, I believe this is an issue with tmux specifically.  I have no issues when I run a naked terminal (tried it in both XTerm (v. 337) and xfce4-terminal v. 0.8.7.4; both were fine alone, but had the same issue in tmux).

Thanks again for working on this feature!  It's a major win.

**Comment by lhca** [comment]
@koreymacdougall - Thanks. I find it very useful myself. Would it be possible for you to update to the latest commit (f45062a) of multiple-days? I see from the screendump that you are at least three weeks behind ;-)

**Comment by koreymacdougall** [comment]
lol whoops.  You're right.  

I've updated.  The newer version does feel better.  The red brackets are quite a nice touch.  I find the default 3-tone coloring of the dates (current/appts/no-appts) more a bit confusing at times. 

The bug does persist, however.  Again, it is limited to tmux; stand-alone terminals work like a charm.

![calcurse_multiple_days](https://user-images.githubusercontent.com/22301892/52583817-86526280-2dfe-11e9-9dd3-73eac0ef27ea.png)

**Comment by lhca** [comment]
> The bug does persist, however. Again, it is limited to tmux; stand-alone terminals work like a charm. So I doubt it's a bug in your code.

It looks like incorrect updating of the screen - a ncurses job. To me ncurses appears rock-solid, so I'm more inclined to think it's a tmux problem.

I have just run calcurse in a tmux session, but didn't see any problem like yours. How many days have you set calcurse to load at a time?

**Comment by lfos** [comment]
Ẃhat's the status of this? Ready for a PR?

**Comment by lhca** [comment]
I think it is. I have used it for several weeks now and am very satisfied with the result. The tmux problem mentioned above is strange. Nothing in calcurse related to the basic screen handling has changed.

There is one undecided issue, though: the number of days in the appointments panel. Currently it is configurable between one and seven with five as the default. I still believe that there should be a limit, but seven is too low. I suggest three weeks (21 days) with seven days as the default.

**Comment by lfos** [comment]
Well, let's not waste time bikeshedding... We can make 21 days an upper limit and still change that later.

**Comment by lhca** [comment]
Right. I have just opened a PR #190.

**Comment by lhca** [comment]
New PR #193.

**Comment by cvoges12** [comment]
Is it possible to hide the days that have no appointments as well?

**Comment by lfos** [comment]
No, they are needed, otherwise you cannot add an item to an empty day.

**Comment by cvoges12** [comment]
@lfos Why not simply use the calendar view to select the day which you will add the appointments? That already exists as a feature.

**Comment by lfos** [comment]
We currently do display empty days, though. Do you suggest do display an empty day if it is selected in the calendar panel? You could browse through the days using the appointments panel, but would skip over empty days. Then, when using the calendar panel, an empty day appears in the appointments panel when you select that day in the calendar panel and disappears once you move to the next day (which may or may not add another dummy item at the same time).

@lhca What do you think about that suggestion?

**Comment by lhca** [comment]
@cvoges12 I miss a (convincing) reason in the first place. Are you using the commits of this PR?

I see it as a bonus and a natural consequence of a new UI. The main attention (of the user) is on the appointments panel, not on the calendar. Without empty days the focus must be changed from the appointments panel to the calendar, the selected day moved to "create" a new day in the appointments panel, and the focus changed back to the appointments panel, before you can insert anything. It's like having a paper calendar with only "busy" pages where you must insert a blank page before adding anything to a "free" day.

By the way, is "none" the proper text to use? Should it be "(empty)"? Or "(free)"? Or just "-"?

**Comment by lfos** [comment]
@lhca Sure, that is one way to see it. We should be open to other suggestions at this point though (not saying that I favor one implementation over another). You can also argue that the appointment panel should be used to browse appointments and the calendar panel should be used to browse the calendar! Frankly, this sounds more "natural" to me than the new UI you mentioned. This would allow you to quickly jump from one appointment to the next which is not easily possible with the current interface.

Also, for people with a very sparse calendar (say only one appointment each week), the multiple days feature with empty days is not really useful as it does not give you more information than the current single day implementation.

**Comment by lhca** [comment]
Hmm. To continue the paper calendar analogy, a person with a sparse calendar will probably not buy one.

I am still in the dark as to @cvoges12´ concerns. If the problem is a lot of empty days in the appointments panel, I would answer: don´t use the possibility of multiple days. With only one day in the panel, the difference from earlier versions are hardly noticable.

But for people with a dense calendar I do believe that multiple days (with empty days) are an improvement. It might be possible to cater for both types of calendar (dense/spare) later, but not now.

**Comment by lfos** [comment]
Here is how I see it: we have a new feature X that we want to implement and two possible implementations (a) and (b). Implementation (a) allows for quickly browsing the calendar from the appointments panel and allows inserting into day empty day without switching panels but is not very useful for sparse calendars. Implementation (b) only lets you browse appointments from the appointments panel (and the calendar from the calendar panel), inserting items to empty days is more cumbersome but the space in appointments panel is used "optimally" in any case.

TL;DR: Implementation (a) is for busy people and implementation (b) is for people with few appointments. Personally, my calendar is usually pretty packed, so I am also leaning slightly towards (a) but that does not mean we should discard (b) right away.

> I am still in the dark as to @cvoges12´ concerns. If the problem is a lot of empty days in the appointments panel, I would answer: don´t use the possibility of multiple days. With only one day in the panel, the difference from earlier versions are hardly noticable.

Well, as a reply to "implementation (b) outperforms implementation (a) is scenario Y", your reply is "I don't care about scenario Y, just don't use the new feature X then". So that is not really a new argument. However, it is true that with the current suggestion, we have to decide for one of the scenarios described above.

> 
> But for people with a dense calendar I do believe that multiple days (with empty days) are an improvement. It might be possible to cater for both types of calendar (dense/spare) later, but not now.

I agree on that, we can focus on getting the current implementation merged and improve it later. The current PR is already complex enough and in a pretty advanced state -- and many of the changes will be required anways, so it is certainly a good starting point.

We might consider other options like implementing both modes and making it configurable or showing all appointments and making empty days magically appear when scrolling through the appointments panel. Just two ideas that came into my mind without thinking too much about it.

**Comment by lfos** [comment]
Support for this has been merged to master. Thanks to @lhca for his amazing work!
